src/main/java/com/mojang/serialization/DataResult.java

```diff
@@ -20,298 +18,354 @@
-    }
+        @Override
+        public <E extends Throwable> R getOrThrow(final Function<String, E> exceptionSupplier) throws E {
+            return value;
+        }
+
+        @Override
+        public <E extends Throwable> R getPartialOrThrow(final Function<String, E> exceptionSupplier) throws E {
+            return value;
+        }
+
+        @Override
+        public <T> DataResult<T> map(final Function<? super R, ? extends T> function) {
+            return new Success<>(function.apply(value), lifecycle);
+        }
+
+        @Override
+        public <T> T mapOrElse(final Function<? super R, ? extends T> successFunction, final Function<? super Error<R>, ? extends T> errorFunction) {
+            return successFunction.apply(value);
+        }
+
+        @Override
+        public DataResult<R> ifSuccess(final Consumer<? super R> ifSuccess) {
+            ifSuccess.accept(value);
+            return this;
+        }
 
-    public static class PartialResult<R> {
-        private final Supplier<String> message;
-        private final Optional<R> partialResult;
+        @Override
+        public DataResult<R> ifError(final Consumer<? super Error<R>> ifError) {
+            return this;
+        }
 
-        public PartialResult(final Supplier<String> message, final Optional<R> partialResult) {
-            this.message = message;
-            this.partialResult = partialResult;
+        @Override
+        public DataResult<R> promotePartial(final Consumer<String> onError) {
+            return this;
         }
 
-        public <R2> PartialResult<R2> map(final Function<? super R, ? extends R2> function) {
-            return new PartialResult<>(message, partialResult.map(function));
+        @Override
+        public <R2> DataResult<R2> flatMap(final Function<? super R, ? extends DataResult<R2>> function) {
+            return function.apply(value).addLifecycle(lifecycle);
         }
 
-        public <R2> PartialResult<R2> flatMap(final Function<R, PartialResult<R2>> function) {
-            if (partialResult.isPresent()) {
-                final PartialResult<R2> result = function.apply(partialResult.get());
-                return new PartialResult<>(() -> appendMessages(message.get(), result.message.get()), result.partialResult);
+        @Override
+        public <R2> DataResult<R2> ap(final DataResult<Function<R, R2>> functionResult) {
+            final Lifecycle combinedLifecycle = lifecycle.add(functionResult.lifecycle());
+            if (functionResult instanceof final Success<Function<R, R2>> funcSuccess) {
+                return new Success<>(funcSuccess.value.apply(value), combinedLifecycle);
+            } else if (functionResult instanceof final Error<Function<R, R2>> funcError) {
+                return new Error<>(funcError.messageSupplier, funcError.partialValue.map(f -> f.apply(value)), combinedLifecycle);
+            } else {
+                throw new UnsupportedOperationException();
```

Comment:
Maybe `TODO` for when [pattern matching](https://openjdk.org/jeps/441) shows up?